amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java

```diff
@@ -0,0 +1,348 @@
+/**
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.streamnative.pulsar.handlers.amqp;
+
+import static org.apache.qpid.server.protocol.ErrorCodes.INTERNAL_ERROR;
+
+import io.streamnative.pulsar.handlers.amqp.admin.AmqpAdmin;
+import io.streamnative.pulsar.handlers.amqp.admin.model.BindingParams;
+import io.streamnative.pulsar.handlers.amqp.admin.model.ExchangeDeclareParams;
+import io.streamnative.pulsar.handlers.amqp.admin.model.QueueDeclareParams;
+import io.streamnative.pulsar.handlers.amqp.common.exception.AoPServiceRuntimeException;
+import io.streamnative.pulsar.handlers.amqp.impl.PersistentExchange;
+import io.streamnative.pulsar.handlers.amqp.impl.PersistentQueue;
+import io.streamnative.pulsar.handlers.amqp.utils.MessageConvertUtils;
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import lombok.extern.log4j.Log4j2;
+import org.apache.pulsar.broker.PulsarServerException;
+import org.apache.pulsar.client.api.Message;
+import org.apache.pulsar.client.api.Producer;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.SubscriptionInitialPosition;
+import org.apache.pulsar.client.api.SubscriptionType;
+import org.apache.pulsar.common.naming.TopicDomain;
+import org.apache.pulsar.common.util.FutureUtil;
+import org.apache.qpid.server.protocol.v0_8.AMQShortString;
+import org.apache.qpid.server.protocol.v0_8.FieldTable;
+import org.apache.qpid.server.protocol.v0_8.transport.AMQMethodBody;
+import org.apache.qpid.server.protocol.v0_8.transport.BasicAckBody;
+import org.apache.qpid.server.protocol.v0_8.transport.BasicConsumeOkBody;
+import org.apache.qpid.server.protocol.v0_8.transport.MessagePublishInfo;
+import org.apache.qpid.server.protocol.v0_8.transport.MethodRegistry;
+import org.apache.qpid.server.protocol.v0_8.transport.QueueDeclareOkBody;
+
+/**
+ * Amqp Channel level method processor.
+ */
+@Log4j2
+public class AmqpMultiBundlesChannel extends AmqpChannel {
```

Comment:
The receiveBasicPublish method needs to be overridden to address default exchange issues such as:
```
public void receiveBasicPublish(AMQShortString exchange, AMQShortString routingKey, boolean mandatory, boolean immediate) {
        if (log.isDebugEnabled()) {
            log.debug("RECV[{}] BasicPublish[exchange: {} routingKey: {} mandatory: {} immediate: {}]", channelId, exchange, routingKey, mandatory, immediate);
        }
        if (isDefaultExchange(exchange)) {
            ExchangeDeclareParams params = new ExchangeDeclareParams();
            params.setType(ExchangeDefaults.DIRECT_EXCHANGE_CLASS);
            params.setInternal(false);
            params.setAutoDelete(false);
            params.setDurable(true);
            params.setPassive(false);
            params.setArguments(new HashMap<>());
            getAmqpAdmin().exchangeDeclare(connection.getNamespaceName().toString(), AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, params)
.thenCompose(__ -> {

                BindingParams bindingParams = new BindingParams();
                bindingParams.setRoutingKey(routingKey.toString());
                bindingParams.setArguments(new HashMap<>());

                return getAmqpAdmin().queueBind(connection.getNamespaceName().toString(), AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, routingKey.toString(), bindingParams).exceptionally(t -> {
                    log.error("Failed to bind queue {} to exchange {}.", routingKey, AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, t);
                    handleAoPException(t);
                    return null;
                });
            }).thenRun(() -> {
                MessagePublishInfo info = new MessagePublishInfo(AMQShortString.valueOf(AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE), immediate, mandatory, routingKey);
                setPublishFrame(info, null);
            }).exceptionally(t -> {
                log.error("Failed to bind queue {} to exchange {}.", routingKey, AbstractAmqpExchange.DEFAULT_EXCHANGE_DURABLE, t);
                handleAoPException(t);
                return null;
            }).join();
        } else {
            MessagePublishInfo info = new MessagePublishInfo(exchange, immediate, mandatory, routingKey);
            setPublishFrame(info, null);
        }
    }
```

Reply:
Yes, I want resolve this case in next PR, I add a TODO comment first.

Reply:
fixed